app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,168 @@
-    batch
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+      if ar_object.is_a?(Account) && !ar_object.should_be_deleted?
+        raise DoNotRetryError, "background deleting account #{ar_object.id} which is not scheduled for deletion"
+      end
 
-  def batch_callbacks(batch)
-    %i[success complete].each { |name| batch.on(name, self.class, callback_options) }
-    yield
-    bid = batch.bid
+      ar_object_str = "#{ar_object.class}-#{ar_object.id}"
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(hierarchy) unless hierarchy.first.is_a?(String)
 
-      ReflectionDestroyer.new(object, reflection, caller_worker_hierarchy).destroy_later
+    started = now
+    while hierarchy.present? && now - started < WORK_TIME_LIMIT_SECONDS
+      Rails.logger.info "Starting background deletion iteration with: #{hierarchy.join(' ')}"
+      hierarchy.concat handle_hierarchy_entry(hierarchy.pop)
     end
+
+    @remaining_hierarchy = hierarchy
+    # can be like this instead of in `after_perform`, the benefit is that we don't use undocumented order of callbacks
+    # return unless hierarchy.present?
+    #
+    # # depending on
+    # unlock(resource: lock_key)
+    # self.class.perform_later(*hierarchy)
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
+  def lock_key
+    first_argument = arguments.first
+    first_argument.is_a?(String) ? first_argument : Random.uuid
   end
 
   private
 
-  def called_from_provider_hierarchy?
-    return unless (tenant_id = object.tenant_id)
-    caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
+  def hierarchy_entries_for(...)
+    self.class.send(:hierarchy_entries_for, ...)
```

Comment:
Could it probably be substituted by `delegate :hierarchy_entries_for, to: :class`, private: true` ? Seems cleaner.

Reply:
`private: true` makes the delegated method private, it doesn't allow you to call a private method. This was also my original thought though.